src/main/java/org/purescript/psi/expression/ExpressionIdentifierReference.kt

```diff
@@ -0,0 +1,79 @@
+package org.purescript.psi.expression
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.LocalQuickFixProvider
+import com.intellij.psi.PsiElementResolveResult.createResults
+import com.intellij.psi.PsiNamedElement
+import com.intellij.psi.PsiReferenceBase
+import com.intellij.psi.ResolveResult
+import com.intellij.psi.util.parents
+import org.purescript.file.ExportedValuesIndex
+import org.purescript.psi.PSValueDeclaration
+
+class ExpressionIdentifierReference(expressionConstructor: PSExpressionIdentifier) :
+    LocalQuickFixProvider,
+    PsiReferenceBase.Poly<PSExpressionIdentifier>(
+        expressionConstructor,
+        expressionConstructor.qualifiedIdentifier.identifier.textRangeInParent,
+        false
+    ) {
+
+    override fun getVariants(): Array<Any> =
+        candidates.toList().toTypedArray()
+
+    override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
+        val name = element.name
+        val declaration = candidates.firstOrNull { it.name == name }
+            ?: return emptyArray()
+        // TODO We may not want to have a PsiPolyVariantReference reference here
+        return if (declaration is PSValueDeclaration) {
+            val module = declaration.module
+            val allValueDeclarations = candidates.filterIsInstance<PSValueDeclaration>()
+                .filter { it.name == name }
+                .filter { it.module == module }
+                .toMutableList()
+            createResults(allValueDeclarations)
```

Comment:
I don't know if it's worth keeping poly variant references for value declarations, but I did not remove any functionality.